replica/database.cc

```diff
@@ -2370,9 +2370,12 @@ future<> database::flush_keyspace_on_all_shards(sharded<database>& sharded_db, s
     });
 }
 
-future<> database::flush_all_tables() {
+future<> database::flush_all_tables(database::wait_for_commitlog_reclaim wait_for_reclaim) {
     // see above
     co_await _commitlog->force_new_active_segment();
+    if (wait_for_reclaim) {
+        co_await _commitlog->wait_for_pending_deletes();
+    }
```

Comment:
@tchaikov  This is called too early.

Forcing a new active segment just "seals" the current active segment and opens a new one.
There could be several inactive commitlog segments now that cannot be recycles because they are pinned by memtables that are not flushed yet.
On after flushing the memtables in the loop below, the inactive commitlog segments may be recycled, as all references to them are released.
And only then we should wait for pending deletes.

This is assuming that all deletions start before last flush() call below resolves.  @elcallio please confirm this.

Reply:
> @tchaikov This is called too early.
> 
> Forcing a new active segment just "seals" the current active segment and opens a new one. 
> There could be several inactive commitlog segments now that cannot be recycles because they are pinned by memtables that are not flushed yet. On after flushing the memtables in the loop below, the inactive commitlog segments may be recycled, as all references to them are released. And only then we should wait for pending deletes.

i see,  it tries to erase the flushed and unused segments, and recycles the erased ones. but if the active segment is not flushed, it won't be recycled.

> 
> This is assuming that all deletions start before last flush() call below resolves. @elcallio please confirm this.

actually Calle has confirmed this in his comment of https://github.com/scylladb/scylladb/pull/16757#issuecomment-1893518269. it's my fault not read it carefully.